src/procedures/utility/build_weighted_matrix.c

```diff
@@ -0,0 +1,475 @@
+	}                                                                          \
+}
+typedef struct
+{
+	const Graph *g;		// Graph
+	AttributeID w;	// Weight Edge
+	RelationID rel;
+} compareContext;
+
+void _selectAttribute
+(
+	bool *z,
+	const uint64_t *x,
+	GrB_Index i,
+	GrB_Index j,
+	const 	compareContext *contx
+)
+{
+	Edge _x;
+	Graph_GetEdge(contx->g, (EdgeID) (*x), &_x);
+	SIValue *v = GraphEntity_GetProperty((GraphEntity *) &_x, contx->w);
+	*z = (v != ATTRIBUTE_NOTFOUND) && (SI_NUMERIC & SI_TYPE(*v));
+}
+void _getAttFromID
+(
+	double *z, 
+	const uint64_t *x, 
+	const compareContext *contx
+)
+{
+	Edge _x, _y;
+	Graph_GetEdge(contx->g, (EdgeID) (*x), &_x);
+	SIValue *v = AttributeSet_Get(*_x.attributes, contx->w);
+	ASSERT(v != ATTRIBUTE_NOTFOUND);
+	int info = SIValue_ToDouble(v, z);
+	ASSERT(info == 1);
+}
+
+#define BWM_COMP(X,Y) SIValue_Compare((X), (Y), NULL) == 1
+#define BWM_INF INFINITY
+void _minGetEdgeID 
+COMPAREEDGES
+
+void _reduceToMatrixMin
+TENSORPICK
+
+#undef BWM_COMP
+#undef BWM_INF
+#define BWM_COMP(X,Y) SIValue_Compare((X), (Y), NULL) == -1
+#define BWM_INF -INFINITY
+void _maxGetEdgeID 
+COMPAREEDGES
+
+void _reduceToMatrixMax
+TENSORPICK
+
+void _reduceToMatrixAny
+(
+	uint64_t *z, 
+	const uint64_t *x
+)
+{
+	if(SCALAR_ENTRY(*x))
+	{
+		*z = *x;
+	}
+	else
+	{
+		GrB_Vector _v = AS_VECTOR(*x);
+		GxB_Iterator i = NULL;
+		GrB_Info info = GxB_Iterator_new(&i);
+		ASSERT(info == GrB_SUCCESS);
+		info = GxB_Vector_Iterator_attach(i, _v, NULL);
+		ASSERT(info == GrB_SUCCESS);
+		info = GxB_Vector_Iterator_seek(i, 0);
+		ASSERT(info == GrB_SUCCESS);
+		EdgeID minID = (EdgeID) GxB_Vector_Iterator_getIndex(i);
+		info = GxB_Iterator_free(&i);
+		ASSERT(info == GrB_SUCCESS);
+		*z = minID;
+	}
+}
+// compose multiple label & relation matrices into a single matrix
+// L = L0 U L1 U ... Lm
+// A = L * (R0 + R1 + ... Rn) * L
+//
+// rows = L's main diagonal
+// in case no labels are specified rows is a dense 1 vector: [1,1,...1]
+GrB_Info Build_Weighted_Matrix
+(
+	GrB_Matrix *A,           // [output] matrix (EdgeIDs)
+	GrB_Matrix *A_w,         // [output] matrix (weights)
+	GrB_Vector *rows,        // [output] filtered rows
+	const Graph *g,          // graph
+	const LabelID *lbls,     // [optional] labels to consider
+	unsigned short n_lbls,   // number of labels
+	const RelationID *rels,  // [optional] relationships to consider
+	unsigned short n_rels,   // number of relationships
+	const AttributeID weight,// attribute to return
+	BWM_reduce strategy,     // Decides how singular returned edge id is picked
+	bool symmetric,          // build a symmetric matrix
+	bool compact            // remove unused row & columns
+) {
+	compareContext contx = {.g = g, .w = weight};
+	GrB_Type contx_type = NULL;
+	GxB_IndexBinaryOp minID_indexOP = NULL;
+	GrB_IndexUnaryOp hasAtt = NULL;
+	GrB_BinaryOp minID = NULL, toMatrixMin = NULL, weightOp = NULL;
+	GrB_UnaryOp toMatrix = NULL;
+	GrB_Scalar theta = NULL;
```

Comment:
One line per variable
align variable name, equal sign and value on the same column